pkg/queue/cluster_queue_best_effort_fifo.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"container/heap"
+	"fmt"
+	"reflect"
+
+	kueue "sigs.k8s.io/kueue/api/v1alpha1"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// ClusterQueueBestEffortFIFO is the implementation for the ClusterQueue for
+// BestEffortFIFO.
+type ClusterQueueBestEffortFIFO struct {
+	*ClusterQueueImpl
+
+	// inadmissibleWorkloads are workloads that have been tried at least once and couldn't be admitted.
+	inadmissibleWorkloads map[string]*workload.Info
+}
+
+var _ ClusterQueue = &ClusterQueueBestEffortFIFO{}
+
+const BestEffortFIFO = kueue.BestEffortFIFO
+
+func newClusterQueueBestEffortFIFO(cq *kueue.ClusterQueue) (ClusterQueue, error) {
+	cqImpl := ClusterQueueImpl{
+		heap: heapImpl{
+			less:  byCreationTime,
+			items: make(map[string]*heapItem),
+		},
+	}
+
+	cqBE := &ClusterQueueBestEffortFIFO{
+		ClusterQueueImpl:      &cqImpl,
+		inadmissibleWorkloads: make(map[string]*workload.Info),
+	}
+
+	cqBE.Update(cq)
+	return cqBE, nil
+}
+
+func (cq *ClusterQueueBestEffortFIFO) PushOrUpdate(w *kueue.QueuedWorkload) {
+	key := workload.Key(w)
+	oldInfo := cq.inadmissibleWorkloads[key]
+	if oldInfo != nil {
+		if isWorkloadUpdated(oldInfo.Obj, w) {
+			delete(cq.inadmissibleWorkloads, key)
+		} else {
+			cq.inadmissibleWorkloads[key] = workload.NewInfo(w)
+			return
+		}
+	}
+
+	(*ClusterQueueImpl).PushOrUpdate(cq.ClusterQueueImpl, w)
+}
+
+func (cq *ClusterQueueBestEffortFIFO) Delete(w *kueue.QueuedWorkload) {
+	delete(cq.inadmissibleWorkloads, workload.Key(w))
+	(*ClusterQueueImpl).Delete(cq.ClusterQueueImpl, w)
+}
+
+func (cq *ClusterQueueBestEffortFIFO) AddInadmissibleIfNotPresent(wInfo *workload.Info) (bool, error) {
+	key := workload.Key(wInfo.Obj)
+	if cq.inadmissibleWorkloads[key] != nil {
+		return false, fmt.Errorf("workload %v is already present in inadmissibleWorkloads", key)
+	}
+
+	if _, exist := cq.heap.Get(wInfo); exist {
```

Comment:
```suggestion
	if item := cq.heap.items[key]; item != nil {
```

just because we are doing this in many other places; once we move the heap to its own pkg, we can introduce the Get function and refactor this everywhere; but for now I would remove the `Get` function you added to the heap.

Reply:
Ok. Deleted the Get function.